xds/internal/resolver/watch_service.go

```diff
@@ -91,11 +91,37 @@ type serviceUpdateWatcher struct {
 }
 
 func (w *serviceUpdateWatcher) handleLDSResp(update xdsresource.ListenerUpdate, err error) {
+	if !w.processLDSResp(update, err) {
+		// Return early if we don't have to watch a RouteConfigName.
+		return
+	}
+
+	w.mu.Lock()
+	if w.rdsCancel != nil {
+		w.rdsCancel()
+	}
+	w.mu.Unlock()
+
+	// WatchRouteConfig could invoke handleRDSResp inline if it cannot find a
+	// matching authority from the bootstrap configuration for the one specified
+	// in the given RDS resource name. Therefore, we cannot call it while still
+	// holding the lock (the acquisition of this lock is the first step in
+	// handleRDSResp).
```

Comment:
Should we maybe consider changing that behavior then?  I.e. invoke it in a goroutine?

Reply:
Ping?